### PR TITLE
JSON Based Saving

### PR DESCRIPTION
**Pull request by lilleswing**
Move to a world where we save the model graph representation as json.

1. TensorGraph subclass is fully describable by the **kwargs to the constructor
2. Save **kwargs to the constructor and the model variables

When loading_from_dir, we now construct the model the exact SAME way with a call to the constructor. then call the built in restore functionality to load the weights.

### CODE REVIEW

**Comment by lilleswing** [comment]
This is a STEP towards mapping layer_name -> tensor_values in the checkpoint files for restoring from eager.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 1]
```diff
@@ -1044,8 +1060,21 @@ def get_pre_q_input(self, input_layer):
     pre_q_name = "%s_pre_q" % layer_name
     return self.layers[pre_q_name]
 
+  @classmethod
+  def load_from_dir(cls, model_dir, restore=True):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tensor_graph.py]: New load from dir method.

**Comment by peastman** [comment]
I need to think about this.  On one hand, moving away from pickling seems like a good idea since it's very brittle.  On the other hand, I worry this could be brittle in other ways.  I worry that people will modify model classes and not realize (or not remember) they also have to update the saving code.

I see you modified all the test cases so instead of calling `TensorGraph.load_from_dir()`, they call it on the specific subclass.  Is that actually needed?  It looks to me like it's only defined on the parent class.  Having to call it on the correct subclass would be a big opportunity for bugs.

Instead of having separate `load_from_dir()` and `load_from_dir_depr()` methods, what about having it just look to see whether `kwargs.json` is present in the directory?  Then it can load the correct format.

**Comment by lilleswing** [comment]
The new load mechanism is a class method not a static method.  That means that the class that the function sees is the class that it is called with.  If it is called with TensorGraph instead of the subclass then the correct sub-class constructor will not be called.

Under the hood load_from_dir checks for the kwargs.json and falls back on legacy already, I can just merge the functions if we don't want to expose the deprecated method.

**Comment by lilleswing** [comment]
The brittleness can be fought somewhat by unit tests.
If we have a unit test for each existing model which saves/loads and checks the results are the same then we won't have any issues.  The test will break and people will update the code.

If the model/layers change and the code version is different then the snapshot version the saved model will not load properly and we will not get an exception.  This is no different than how it works now though.

**Comment by peastman** [comment]
>  If it is called with TensorGraph instead of the subclass then the correct sub-class constructor will not be called.

If you store the fully qualified name of the class (both `cls.__module__` and `cls.__name__`) in the json file, then you can determine the correct class to instantiate automatically.

> If we have a unit test for each existing model which saves/loads and checks the results are the same then we won't have any issues.

If someone adds a new attribute to a class but doesn't update the code to save it, the test won't break (because it won't know to check the value of the new attribute).  We'll need to be very diligent about checking for this when reviewing PRs.

**Comment by lilleswing** [comment]
> If you store the fully qualified name of the class (both cls.__module__ and cls.__name__) in the json 
> file, then you can determine the correct class to instantiate automatically.

Yeah that is slick.  I implemented it.  It doesn't work if you have an anonomys class, but we can fall back in that case.

**Comment by lilleswing** [comment]
This PR also has been updated to work with tf.eager in 1.6 now, and I think it is ready to go.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 2]
```diff
@@ -1044,8 +1042,28 @@ def get_pre_q_input(self, input_layer):
...
+    kwargs_file = os.path.join(model_dir, 'kwargs.json')
+    if not os.path.exists(kwargs_file):
+      return TensorGraph.load_from_dir(model_dir)
```
Comment (peastman) [inline deepchem/models/tensorgraph/tensor_graph.py]: Isn't this going to produce an infinite recursion?

**Comment by peastman** [comment]
This is based on the assumption that every model is fully defined by the arguments passed to its constructor.  Currently there are lots of situations where that isn't the case.  Here are a few examples.

- You create a TensorGraph object and manually add layers to it.  That will no longer be savable.
- The constructor always creates an Adam optimizer, but you can change that later by calling `set_optimizer()`.  That won't get saved.
- You want to train part of the model separately from the rest, so you call `create_submodel()`.  That won't get saved.

There probably are many others.

**Comment by rbharath** [comment]
I'm going to close this old PR since we've moved to using Keras to save models. Feel free to re-open if we should still consider this PR.

**Pull request closed without merging**